File: 配套资源/配套代码/程序设计竞赛专题挑战教程-代码/第3章汉诺塔.py
#汉诺塔 hlanqiaoOJ 1512

def hanoi(x, y, z, n):
    global sum
    if (n == 1):
        sum += 1
        if (sum == m):  print(f"#{n}: {x}->{z}")
    else:
        hanoi(x, z, y, n-1)
        sum += 1
        if sum == m:    print(f"#{n}: {x}->{z}")
        hanoi(y, x, z, n-1)
n, m = map(int, input().split())
sum = 0
hanoi('A', 'B', 'C', n)
print(sum)

File: 配套资源/配套代码/程序设计竞赛专题挑战教程-代码/第3章汉诺塔.cpp
//汉诺塔 lanqiaoOJ 1512 

#include<bits/stdc++.h>
using namespace std;
int sum = 0, m;
void hanoi(char a,char b,char c,int n){ //递归函数
    if(n==1) {
        sum++;
        if(sum==m) cout<<"#"<<n<<": "<<a<<"->"<<c<<endl;
    }
    else {                              //把n层递归到n-1层
        hanoi(a,c,b,n-1);               //
        sum++;
        if(sum==m) cout<<"#"<<n<<": "<<a<<"->"<<c<<endl;
        hanoi(b,a,c,n-1);
    }
}
int main(){
    int n;    cin>>n>>m;
    hanoi('A','B','C',n);
    cout<<sum<<endl;
    return 0;
}
